proof/refine/ARM/Invariants_H.thy

```diff
@@ -260,6 +260,30 @@ abbreviation tcbs_of' :: "kernel_state \<Rightarrow> obj_ref \<Rightarrow> tcb o
 abbreviation tcb_scs_of' :: "kernel_state \<Rightarrow> obj_ref \<Rightarrow> obj_ref option" where
   "tcb_scs_of' s \<equiv> tcbs_of' s |> tcbSchedContext"
 
+abbreviation scTCBs_of :: "kernel_state \<Rightarrow> obj_ref \<Rightarrow> obj_ref option" where
+  "scTCBs_of s \<equiv> scs_of' s |> scTCB"
+
+definition sqheap_refs_retract_at :: "('a \<rightharpoonup> 'b) \<Rightarrow> ('b \<rightharpoonup> 'a) \<Rightarrow> 'a \<Rightarrow> bool" where
+  "sqheap_refs_retract_at heap symheap p \<equiv> \<forall>r. pred_map_eq r heap p \<longrightarrow> pred_map_eq p symheap r"
+
+definition sqheap_refs_retract :: "('a \<rightharpoonup> 'b) \<Rightarrow> ('b \<rightharpoonup> 'a) \<Rightarrow> bool" where
+  "sqheap_refs_retract heap symheap \<equiv> \<forall>p. sqheap_refs_retract_at heap symheap p"
+
+definition sqheap_refs_inv :: "('a \<rightharpoonup> 'b) \<Rightarrow> ('b \<rightharpoonup> 'a) \<Rightarrow> bool" where
+  "sqheap_refs_inv heap symheap \<equiv> sqheap_refs_retract heap symheap \<and> sqheap_refs_retract symheap heap"
+
+lemmas sqheap_refs_inv_defs = sqheap_refs_inv_def sqheap_refs_retract_def sqheap_refs_retract_at_def
+
+lemma sqheap_refs_inv_def2:
+  "sqheap_refs_inv heap symheap \<equiv> \<forall>p q. pred_map_eq q heap p \<longleftrightarrow> pred_map_eq p symheap q"
+  by (auto simp: atomize_eq sqheap_refs_inv_defs)
+
+abbreviation tcbs_scs_sym_refs where
+  "tcbs_scs_sym_refs s \<equiv> sqheap_refs_inv (tcb_scs_of' s) (scTCBs_of s)"
+
+abbreviation replies_scs_sym_refs where
+  "replies_scs_sym_refs s \<equiv> sqheap_refs_inv (scReplies_of s) (replySCs_of s)"
```

Comment:
They are the same.

Reply:
...except for the usual disagreements over definitions vs abbreviations, etc. :-)

Reply:
I think the difference goes far beyond that, and in my opinion what we have here as well as the related part of at the end of AInvs have excessive layers of definitions that we don’t benefit from and could do without.

The heap symmetry is a simple notion. It is perfectly fine to define it directly. Generally speaking, a definition (or abbreviation for that matter) should capture the concept succinctly, without coercing anyone into a particular way of thinking when there are other ways to arrive at that concept. Coercing people in that way will be counterproductive because it will impose extra mental load on them.

If some of the subconcepts are useful, it can be derived or given as alternatives, rather than forced building blocks.

But I also would like to say that I don’t see much point in redefining simple things such as forall and implication under different names. These are simple enough things that it is easy to capture the meaning just by glancing the statements. Why do we want to encapsulate and rename them to obscure their meaning? I don’t see the point of having `pred_map_eq` in the heap context to start with. It is equivalent to `m x = Some y` and, to be honest, I don’t quite understand why one would want to hide that.

Reply:
Well, there might be some cases where a certain combination of simple components is very much recurrent and worthwhile to package it up, but I don’t think that is the case for what we have here. This is quite an overkill for a fairy simple concept and we could do with some streamlining.

Reply:
I didn't know that `sym_heap` existed, so I'm happy to look into use that. However, please keep in mind that I didn't pull these definitions out of thin air, I copied them from DetSchedSchedule so they are consistent with much of the existing proofs.

Reply:
The point of making definitions is to control rewriting, and to build abstractions together with machinery that works at a higher level. It's very possible I failed to do that with the definitions that Mitch copied here. But in general, it's not sustainable to expose every low-level logical connective or quantifier to the simplifier and classical prover, so I would often prefer even a flawed attempt at building abstractions over exposing everything to the simplifier.

Reply:
From what I have seen so far, `sym_heap` and projections play quite nicely with simplifier/automation. The only thing you might want to keep in mind is to remove `opt_mapE` from the rule set. That is usually the culprit when the simplifier is too aggressive.

Frankly, I am skeptical of the gain in choosing automation friendly definition at the cost of human brain efficiency for these particular constructs. Also, I really struggle to understand why you’d want to hide away projections under definitions when we have been moving towards using projections more extensively.

Reply:
I also have a comment from another angle. There is version 1 in AInvs, which is not the same thing as `sym_heap`. So, `sym_heap` was added, serving its purpose perfectly well. Then, now you changed it to version 2 to be the same thing as `sym_heap`. Add this will calls for some considerations on what to do with the version 1 and `sym_heap`, I would think.  So, could you tell me your plan on this if you are pushing for merging this version 2  (Mitch and Matt?) ?

Reply:
I'm not pushing for merging this version 2. I said above that I did not know that `sym_heap` existed and that I would look into using it. I have already partially updated this PR to use `sym_heap` (not pushed yet). If you are concerned about using `pred_map_eq` then we can have further discussion, but that is not something new that I am trying to introduce and so I did not consider that there was any problem using it.

Reply:
@mktnk3, since version 1 came before `sym_heap`, can you tell us what your plan was? I know that the stuff in AInvs is far from ideal, so I'd be happy if you have ways to improve it. W.r.t to this PR, I am ok with what @MitchellBuckley has proposed, but I'm also happy for you to give him some clear direction on what changes you'd like.

Reply:
Version 1 as is does not achieve what `sym_heap` does. They are not the same in purpose, and they are defined totally different way (`sym_heap` is much simpler). So, I added `sym_heap` as something new and different. I have always had in mind that we need some cleanup in AInvs DetSched files, but I consider it as a separate thing.

But now, if this version 2 is the same in terms of semantics as `sym_heap`, and is a modification or improvement of version 1, that would call for some justification/adjustment, is what I meant. You can’t be just adding another thing on top of these, I’d say.

Reply:
Ok, we can see how Mitch goes with the change for now, I suppose.

Reply:
To be honest, I don’t see the need of introducing`pred_map_eq`, either. It not used in Refine so far, except for the very few places in some proofs for AInvs invariants. We have been doing fine with all the Refine proper stuff without using this definition, and we do use the projection extensively.